lib/modules/manager/pip-compile/artifacts.spec.ts

```diff
@@ -330,6 +341,63 @@ describe('modules/manager/pip-compile/artifacts', () => {
     });
   });
 
+  describe('getRegistryUrlFlagsFromPackageFile()', () => {
+    it('handles both registryUrls and additionalRegistryUrls', () => {
+      hostRules.find.mockReturnValue({});
+      expect(
+        getRegistryUrlVarsFromPackageFile({
+          deps: [],
+          registryUrls: ['https://example.com/pypi/simple'],
+          additionalRegistryUrls: ['https://example2.com/pypi/simple'],
+        }),
+      ).toEqual({
+        PIP_INDEX_URL: 'https://example.com/pypi/simple',
+        PIP_EXTRA_INDEX_URL: 'https://example2.com/pypi/simple',
+      });
+    });
+
+    it('handles multiple additionalRegistryUrls', () => {
+      hostRules.find.mockReturnValue({});
+      expect(
+        getRegistryUrlVarsFromPackageFile({
+          deps: [],
+          additionalRegistryUrls: [
+            'https://example.com/pypi/simple',
+            'https://example2.com/pypi/simple',
+          ],
+        }),
+      ).toEqual({
+        PIP_EXTRA_INDEX_URL:
+          'https://example.com/pypi/simple https://example2.com/pypi/simple',
+      });
+    });
+
+    it('uses extra index URLs with no auth', () => {
+      hostRules.find.mockReturnValue({});
+      expect(
+        getRegistryUrlVarsFromPackageFile({
+          deps: [],
+          registryUrls: ['https://example.com/pypi/simple'],
+        }),
+      ).toEqual({ PIP_INDEX_URL: 'https://example.com/pypi/simple' });
+    });
+
+    it('uses auth from extra index URLs matching host rules', () => {
+      hostRules.find.mockReturnValue({
+        username: 'user',
+        password: 'password',
+      });
+      expect(
+        getRegistryUrlVarsFromPackageFile({
+          deps: [],
+          registryUrls: ['https://example.com/pypi/simple'],
+        }),
+      ).toEqual({
+        PIP_INDEX_URL: 'https://user:password@example.com/pypi/simple',
```

Comment:
are you sure this isn't written to the compiled file? why we can extract them when they're not written?

Reply:
Only the URL part is in the .in file (i.e. `--index-url https://example.com/pypi/simple`).  The credentials are pulled from the `hostRules`.  They may or may not end up being written to the lockfile, depending on whether or not the `--no-emit-index-url` flag is given.

Reply:
we probably should set `--no-emit-index-url` when adding credentials to env, otherwise we could upload credentials inside lockfile. That should never happen.

@rarkins @secustor WDYT?

Reply:
I agree let's set `--no-emit-index-url` always. If requested, we can make this optional for self-hosted users later on.

Reply:
I've been looking into other methods of passing credentials out to pip-compile so that they won't appear in the emitted index URLs, even when `--no-emit-index-url` isn't given.
The most promising to me is to use the Keyring module (https://pypi.org/project/keyring/).  It would be pretty simple to write a backend for that module that pulls credentials from environment variables, so that they don't have to be passed directly in index URLs.  I'm pretty sure that would require making changes to the `pip-tools` tool in containerbase to install `keyring` and whatever keyring backend I built alongside `pip-compile`.  That's a bit heavy-handed for just this one purpose.  @viceice do you think that's worth it?

Reply:
we could try it later. let's doing something more simple for now.

- save current command
- run pip-compile
- replace generated command with old command

WDYT?

Reply:
If Renovate passes `--no-emit-index-url` I think the header should reflect that so that people aren't confused if a locally-generated file doesn't match what Renovate generates.  I'm OK with passing `--no-emit-index-url` any time credentials are passed for now.  Once my changes in containerbase/base#2145 are in I can open another PR to change over to that.

Reply:
@viceice I made the changes to pass `--no-emit-index-url` whenever credentials are present in any index URLs.